### ISSUE TITLE
[Feature Request] Support LocationLink for PeekDefinition

### ISSUE THREAD

**Issue by clason** [prabirshrestha/vim-lsp#474]
The [LSP specification](https://microsoft.github.io/language-server-protocol/specification#textDocument_definition) for `Goto Definition` allows for providing a `LocationLink` that includes a `targetRange` for the structure enclosing the linked symbol (for example, the function body corresponding to the definition). This could be used by `PeekDefinition` to restrict the popup to exactly the relevant function (assuming it is short enough); there are also languages (such as TeX) where the targetted symbol does not (necessarily) occur on the first line of this range, so currently the popup is missing relevant information.

To support this, vim-lsp should set the `ClientCapabilties.textDocument.definition.linkSupport` field to `true`. An example response using the `LocationLink` (from https://github.com/latex-lsp/texlab) is
```json
{"jsonrpc":"2.0","result":[{"originSelectionRange":{"end":{"character":11,"line":0},"start":{"character":6,"line":0}},"targetRange":{"end":{"character":13,"line":3},"start":{"character":0,"line":1}},"targetSelectionRange":{"end":{"character":38,"line":1},"start":{"character":33,"line":1}},"targetUri":"file:///D:/latex-lsp/texlab-dev/issue/test1.tex"}],"id":27}
```
(the `targetRange` is the one the popup should show; the `targetSelectionRange` is the actual symbol that should be highlighted).

**Comment by thomasfaingnaert** [comment]
In reference to https://github.com/prabirshrestha/vim-lsp/issues/486#issuecomment-529497969.
This should be fairly easier than #486 to implement :)
One thing to keep in mind is that it needs to work for three separate cases:

1. Preview window
2. Vim popups
3. Neovim floats

Other than that, I would probably:
1. Set the correct capabilities and check the response from the language server manually in the log, just to make sure everything works there
2. Implement displaying a range of lines in all 3 cases above (the most work :(, mainly in `output.vim`). I would start with a hardcoded range of lines to test with.
3. Figure out how to get the information from the language server to `output.vim`.

**Comment by clason** [comment]
@thomasfaingnaert 
> This should be fairly easier than #486 to implement :)

That's what I was hoping for :)

> One thing to keep in mind is that it needs to work for three separate cases:

Thanks, I'll keep that in mind!

> Other than that, I would probably:

Yes, that much is clear; I was fishing for a bit more low-level hints ("you need to look at these functions"; "check out this merged PR that did something similar") since I'm not familiar with the vim-lsp code base at all.

**Comment by clason** [comment]
OK, here is my plan:

1. Deal with the capabilities *(to be replaced by actual plan)*
2. Write a new `lsp#ui#vim#utils#locationlinks_to_loc_list` that parses a list of `LocationLink`s to obtain a *final* line (and col) in addition to a starting line and col.
3. Write a new  `handle_locationlink` that calls this in place of `handle_location`. Instead of passing the full file buffer to `lsp#ui#vim#output#preview`, it uses the additional information to only pass the specified lines (`readline(...)[startline,stopline]`?). This would eliminate the need to set cursor position, too.
4. Adapt `lsp#ui#vim#definition` to use `handle_locationlink` if the capability is set.
5. Repeat for `Implementation` etc.

Does that sound reasonable?

**Comment by thomasfaingnaert** [comment]
> Although looking at it, these functions all seem to already get passed the data, which includes the preview content -- don't I need to rather adapt the function that evaluates the `PeekDefinition` response to extract this content from the actual buffer? Where would that be?

This would fall under step 3 in my simplistic step-by-step plan. I would start by expanding `output.vim` to allow setting a range. The main entrypoint is `lsp#ui#vim#output#preview`: I would probably add a key to options that would allow passing a range instead of a line.

Or even simpler: add an option to set the height of the popup. That way, we set `cursor` to the first line of the location link, and `height` to `last_line - first_line + 1`. This will waste screenspace though, which may/may not be what you want. For example, you may want to be able to focus the preview and scroll around, for which a larger preview is useful.

You would then pass the correct `cursor` and `height` here: https://github.com/prabirshrestha/vim-lsp/blob/e7480252ecd7a5084563c4027c097132f6f84f08/autoload/lsp/ui/vim.vim#L490-L495.

For the response parsing: https://github.com/prabirshrestha/vim-lsp/blob/e7480252ecd7a5084563c4027c097132f6f84f08/autoload/lsp/ui/vim.vim#L92-L99 sends the request. We can see from that line that `s:handle_location` will be called. This is where the response is parsed. The `data` parameter contains all necessary info, and is passed to https://github.com/prabirshrestha/vim-lsp/blob/e7480252ecd7a5084563c4027c097132f6f84f08/autoload/lsp/ui/vim/utils.vim#L1.

> EDIT For 3. it looks like I first need to write a new lsp#ui#vim#utils#locationlinks_to_loc_list function that gets called in place of locations_to_loc_list if LocationLink capabilities are detected, right?

This is indeed where the parsing should be done. I just checked the spec, and the regular `Location` also has an `end`, but it's not used in vim-lsp. Might be worth checking first if texlab returns the correct end using regular locations, then you wouldn't need to completely implement locationlink.

**Comment by thomasfaingnaert** [comment]
>     2. Write a new `lsp#ui#vim#utils#locationlinks_to_loc_list` that parses a list of `LocationLink`s to obtain a _final_ line (and col) in addition to a starting line and col.
The logic will probably logic very similar to `lsp#ui#vim#utils#locations_to_loc_list`, so can also be in the same function. We can change that later, though, let's not worry about that.

>     3. Write a new  `handle_locationlink` that calls this in place of `handle_location`. Instead of passing the full file buffer to `lsp#ui#vim#output#preview`, it uses the additional information to only pass the specified lines (`readline(...)[startline,stopline]`?). This would eliminate the need to set cursor position, too.
True, but then you lose the ability to focus the popup and scroll if you want some more context.

>     4. Adapt `lsp#ui#vim#definition` to use `handle_locationlink` if the capability is set. _(see 1.)_
I think we better do this at the response side: only have one response callback that detects if it gets locations or locationlinks. You never know a server might send locations even though we advertise locationlink support.

Sounds good.
Hopefully this reponse is a little bit clearer. If you have any further questions, let me know!

**Comment by clason** [comment]
> The logic will probably logic very similar to lsp#ui#vim#utils#locations_to_loc_list, so can also be in the same function. We can change that later, though, let's not worry about that.

Yes, I was planning on ~stealing~adapting heavily from that. My plan was indeed to start as a separate function to avoid breaking existing functionality as much as possible and then refactor later. (That was also one motivation for `handle_locationlink`.)

> Or even simpler: add an option to set the height of the popup. That way, we set cursor to the first line of the location link, and height to last_line - first_line + 1. This will waste screenspace though, which may/may not be what you want. For example, you may want to be able to focus the preview and scroll around, for which a larger preview is useful.

Ah, I was not thinking about that possible use case -- for me the Peek was just a static preview to see what the symbol points to. That kinda conflicts with my plan above; I'll have to think about that.

You're right that in this case, adding `height` to the preview call would probably be better. (Although my approach would not need changes to the float positioning...)

> I think we better do this at the response side: only have one response callback that detects if it gets locations or locationlinks. You never know a server might send locations even though we advertise locationlink support.

So basically check inside the `handle_location` function whether it got passed a `location` or a `locationLink` and then branch accordingly? That might be easier than having to check capabilities.

> Hopefully this reponse is a little bit clearer. If you have any further questions, let me know!

Yes, that was helpful, thank you!

**Comment by thomasfaingnaert** [comment]
> > Or even simpler: add an option to set the height of the popup. That way, we set cursor to the first line of the location link, and height to last_line - first_line + 1. This will waste screenspace though, which may/may not be what you want. For example, you may want to be able to focus the preview and scroll around, for which a larger preview is useful.
> 
> Ah, I was not thinking about that possible use case -- for me the Peek was just a static preview to see what the symbol points to. That kinda conflicts with my plan above; I'll have to think about that.
> 
> You're right that in this case, adding `height` to the preview call would probably be better. (Although my approach would not need changes to the float positioning...)

Changing the float positioning is indeed not a very pleasant experience, trust me :p 
Honestly, it may not be worth it. For the moment, I think using you solution by limiting the buffer content is fine.

> > I think we better do this at the response side: only have one response callback that detects if it gets locations or locationlinks. You never know a server might send locations even though we advertise locationlink support.
> 
> So basically check inside the `handle_location` function whether it got passed a `location` or a `locationLink` and then branch accordingly? That might be easier than having to check capabilities.

Precisely. In an ideal world, this would not be necessary. Unfortunately, it seems not all clients adhere as strictly to the LSP protocol as I would like, but "it's fine because it works in VSCode". I don't expect this to be a problem, but better safe than sorry, I guess.

**Comment by thomasfaingnaert** [comment]
If you want to go the `height` route without doing too much work: you could set `g:lsp_preview_max_height` to the correct height and then reset it afterwards instead of adding an option. Whether that's an ugly hack or a brilliant workaround, I leave up to you.

**Comment by clason** [comment]
> Changing the float positioning is indeed not a very pleasant experience, trust me :p
Honestly, it may not be worth it. For the moment, I think using you solution by limiting the buffer content is fine.

I'll limit this to the `LocationLink` code path, then -- assuming that if such a range is provided, that's all people want to `Peek` at (otherwise there's still the `Goto`). I think I have a broad idea of how to go about this now.

> If you want to go the height route without doing too much work: you could set g:lsp_preview_max_height to the correct height and then reset it afterwards instead of adding an option. Whether that's an ugly hack or a brilliant workaround, I leave up to you.

That's a nice idea (code reuse 4tw!), I'll play around with that. (I was thinking about how this would interact with `lsp_preview_max_height`; this way would make it trivial :))

> Precisely. In an ideal world, this would not be necessary. Unfortunately, it seems not all clients adhere as strictly to the LSP protocol as I would like, but "it's fine because it works in VSCode". I don't expect this to be a problem, but better safe than sorry, I guess.

Yeah, texlab is strictly by the book, which irritated me a bit at first since it broke some previously working behavior (you might remember...), but I am starting to see the point of it.

**Comment by clason** [comment]
@thomasfaingnaert Well, that was easier than expected :) #498 

Only works in neovim, though; not sure why vim shows the wrong viewport.
